test/unit/test_environment.py

```diff
@@ -0,0 +1,160 @@
+
+    assert environment.read_hyperparameters() == ALL_HPS
+
+
+def test_split_hyperparameters_only_provided_by_user():
+    assert environment.split_hyperparameters(USER_HPS) == ({}, USER_HPS)
+
+
+def test_split_hyperparameters_only_provided_by_sagemaker():
+    assert environment.split_hyperparameters(SAGEMAKER_HPS) == (SAGEMAKER_HPS, {})
+
+
+def test_split_hyperparameters():
+    assert environment.split_hyperparameters(ALL_HPS) == (SAGEMAKER_HPS, USER_HPS)
+
+
+def test_resource_config(input_config_path):
+    write_json('resourceconfig.json', RESOURCE_CONFIG, input_config_path)
+
+    assert environment.read_resource_config() == RESOURCE_CONFIG
+
+
+def test_input_data_config(input_config_path):
+    write_json('inputdataconfig.json', INPUT_DATA_CONFIG, input_config_path)
+
+    assert environment.read_input_data_config() == INPUT_DATA_CONFIG
+
+
+def test_channel_input_dirs(input_data_path):
+    assert environment.channel_path('evaluation') == str(input_data_path.join('evaluation'))
+    assert environment.channel_path('training') == str(input_data_path.join('training'))
+
+
+@patch('subprocess.check_output', lambda s: b('GPU 0\nGPU 1'))
+def test_gpu_count_in_gpu_instance():
+    assert environment.gpu_count() == 2
+
+
+def test_gpu_count_in_cpu_instance():
+    assert environment.gpu_count() == 0
+
+
+@patch('multiprocessing.cpu_count', lambda: 2)
+def test_cpu_count():
+    assert environment.cpu_count() == 2
+
+
+@patch('sagemaker_container_support.environment.read_resource_config', lambda: RESOURCE_CONFIG)
+@patch('sagemaker_container_support.environment.read_input_data_config', lambda: INPUT_DATA_CONFIG)
+@patch('sagemaker_container_support.environment.read_hyperparameters', lambda: ALL_HPS)
+@patch('sagemaker_container_support.environment.cpu_count', lambda: 8)
+@patch('sagemaker_container_support.environment.gpu_count', lambda: 4)
+def test_environment_create():
+    env = environment.Environment.create(session=Mock())
+
+    assert env.num_gpu == 4
+    assert env.num_cpu == 8
+    assert env.input_dir == '/opt/ml/input'
+    assert env.input_config_dir == '/opt/ml/input/config'
+    assert env.model_dir == '/opt/ml/model'
+    assert env.output_dir == '/opt/ml/output'
+    assert env.hyperparameters == USER_HPS
+    assert env.resource_config == RESOURCE_CONFIG
+    assert env.input_data_config == INPUT_DATA_CONFIG
+    assert env.output_data_dir == '/opt/ml/output/data'
+    assert env.hosts == RESOURCE_CONFIG['hosts']
+    assert env.channel_input_dirs['train'] == '/opt/ml/input/data/train'
+    assert env.channel_input_dirs['validation'] == '/opt/ml/input/data/validation'
+    assert env.current_host == RESOURCE_CONFIG['current_host']
+    assert env.module_name == 'main.py'
+    assert env.module_dir == 'imagenet'
+    assert env.enable_metrics
+    assert env.log_level == logging.WARNING
+
+
+def write_json(name, data, file_path):
+    file_path = file_path.join(name)
```

Comment:
I know this is test code but we should try to have the same bar too. Why can't this function be 

```python
write_json_file(file, data):
```

no need to do that join inside this function. Also no need to return anything if you are already receiving the full path to the file as an argument.

Also, if you are doing a join it should be os.path.join()  to make sure this works on windows too.

Reply:
Good point:

What about ```json_dump```? I believe is more pythonic than ```write_json_file```. 

All the naming in this project so far was strong based on python standard lib  naming conventions and glossary: https://docs.python.org/3/library/ and https://docs.python.org/3/glossary.html

Naming is hard, so let's sync up offline to see if this strategy is good for us.

file_path is not a string but a py.path.local. Which is pytest default way to create temp files: https://docs.pytest.org/en/latest/tmpdir.html

I refactored the functions to the following:
```python
def json_dump(data, path_obj):  # type: (object, py.path.local) -> None
    """Writes JSON serialized data to the local file system path

    Args:
        data (object): object to be serialized
        path_obj (py.path.local): path.local object of the file to be written
    """
    path_obj.write(json.dumps(data))
```

And the tests look like this now:
```python
def test_read_hyperparameters(input_config_path):
    json_dump(ALL_HPS, input_config_path.join('hyperparameters.json'))

    assert environment.read_hyperparameters() == ALL_HPS
```

Let me know your thoughts.